tracing-core/src/callsite.rs

```diff
@@ -84,6 +45,15 @@ pub trait Callsite: Sync {
     ///
     /// [metadata]: ../metadata/struct.Metadata.html
     fn metadata(&self) -> &Metadata<'_>;
+
+    /// Returns the callsite's [`Registration`] in the global callsite registry.
+    ///
+    /// Every type implementing `Callsite` must own a single [`Registration`] which
+    /// is constructed with a static reference to that callsite. The `Registration` is
+    /// used to store the callsite in the global list of callsites. `Registration`s must be
+    /// unique per callsite; a callsite's `Registration` method must always return the
+    /// same `Registration`.
+    fn registration(&'static self) -> &'static Registration;
```

Comment:
It occurs to me that the cyclic dependency between `Callsite` impls and `Registration`s is really only necessary because of this method. I think we can get rid of the need for `Callsite` impls to store `Registration`s that point back at themselves if we just changed the `register` function to take an `&'static Registration` rather than a `&'static dyn Callsite`, and removed `Callsite::registration`.